datafusion/physical-optimizer/src/reverse_order.rs

```diff
@@ -0,0 +1,1446 @@
+        // Can reverse! Apply reversal
+        let reversed_input = sort_input.rewrite(&mut ReverseRewriter).unwrap().data;
+
+        // Check if reversed input satisfies the sort requirement
+        if reversed_input
+            .equivalence_properties()
+            .ordering_satisfy(sort_exec.expr().clone())?
+        {
+            // Check if this is a single-partition DataSourceExec with reverse_scan enabled
+            // In that case, the limit is already handled internally by ReversedParquetStreamWithLimit
+            if is_single_partition_reverse_scan_datasource(&reversed_input) {
+                let total_fetch = limit_exec.skip() + limit_exec.fetch().unwrap_or(0);
+
+                if let Some(with_fetch) = reversed_input.with_fetch(Some(total_fetch)) {
+                    if limit_exec.skip() > 0 {
+                        return Ok(Transformed::yes(Arc::new(GlobalLimitExec::new(
+                            with_fetch,
+                            limit_exec.skip(),
+                            limit_exec.fetch(),
+                        ))));
+                    } else {
+                        return Ok(Transformed::yes(with_fetch));
+                    }
+                }
+
+                return Ok(Transformed::yes(Arc::new(GlobalLimitExec::new(
+                    reversed_input,
+                    limit_exec.skip(),
+                    limit_exec.fetch(),
+                ))));
+            }
+
+            // Otherwise, remove sort but keep limit with reversed input
+            return Ok(Transformed::yes(Arc::new(GlobalLimitExec::new(
+                reversed_input,
+                limit_exec.skip(),
+                limit_exec.fetch(),
+            ))));
+        }
+    }
+
+    // Can't optimize, return original pattern
+    Ok(Transformed::no(Arc::new(GlobalLimitExec::new(
+        Arc::new(sort_exec.clone()),
+        limit_exec.skip(),
+        limit_exec.fetch(),
+    ))))
+}
+
+/// Check if the plan is a single-partition DataSourceExec with reverse_scan enabled
+fn is_single_partition_reverse_scan_datasource(plan: &Arc<dyn ExecutionPlan>) -> bool {
+    // Only optimize for single partition
+    if plan.output_partitioning().partition_count() != 1 {
+        return false;
+    }
+
+    if let Some(data_source_exec) = plan.as_any().downcast_ref::<DataSourceExec>() {
+        if let Some(scan_config) = data_source_exec
+            .data_source()
+            .as_any()
+            .downcast_ref::<FileScanConfig>()
+        {
+            if let Some(parquet_source) = scan_config
+                .file_source
+                .as_any()
+                .downcast_ref::<ParquetSource>()
+            {
+                return parquet_source.reverse_scan();
+            }
+        }
+    }
+    false
+}
+
+/// Remove unnecessary sort based on the logic from EnforceSorting::analyze_immediate_sort_removal
+fn remove_unnecessary_sort(
```

Comment:
Note, i add this to reverse order because after reverse order, we can optimize more to remove the sort, so we don't need to execute the enforce sort optimization again after this optimization.

Reply:
@alamb This is a good point, we can do this after this PR, and split the enforce_sorting to: 

EnforceSortingRequirements and OptimizeSorts.